dart/optimizer/Function.h

```diff
@@ -49,22 +54,145 @@ class Function
 {
 public:
   /// \brief Constructor
-  explicit Function();
+  explicit Function(const std::string& _name = "function");
 
   /// \brief Destructor
   virtual ~Function();
 
+  /// \brief Set the name of this Function
+  virtual void setName(const std::string& _newName);
+
+  /// \brief Get the name of this Function
+  const std::string& getName() const;
+
+  /// \brief Evaluate and return the objective function at the point x.
+  ///
+  /// Note: This function is deprecated in favor of the const Eigen::VectorXd&
+  /// version
+  DEPRECATED(5.1)
+  virtual double eval(Eigen::Map<const Eigen::VectorXd>& _x);
+
   /// \brief Evaluate and return the objective function at the point x
-  virtual double eval(Eigen::Map<const Eigen::VectorXd>& _x) = 0;
+  virtual double eval(const Eigen::VectorXd& _x);
 
   /// \brief Evaluate and return the objective function at the point x
+  ///
+  /// Note: This function is deprecated in favor of the const Eigen::VectorXd&
+  /// version
+  DEPRECATED(5.1)
   virtual void evalGradient(Eigen::Map<const Eigen::VectorXd>& _x,
                             Eigen::Map<Eigen::VectorXd> _grad);
 
   /// \brief Evaluate and return the objective function at the point x
+  virtual void evalGradient(const Eigen::VectorXd& _x,
+                            Eigen::Map<Eigen::VectorXd> _grad);
```

Comment:
We might be able to provide `evalGradient(~)` that takes `Eigen::VectorXd&` for gradient as:
```cpp
void evalGradient(const Eigen::VectorXd& _x, Eigen::VectorXd& _grad)
{
  Eigen::Map<Eigen::VectorXd> tmpGrad(_grad.data(), _grad.size());
  evalGradient(_x, tmpGrad);
}
```
for the users who are not familiar with `Eigen::Map`.